Dispatch an event to define metadata inside the FigureBuilder

m-vo opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | -
| Docs PR or issue | todo

This introduces a `DefineMetadataEvent` that gets dispatched inside the `FigureBuilder` as soon as metadata is requested. It is run *after* the `FigureBuilder` logic and allows to remove/replace/add `Metadata`.

m-vo committed:
inject event dispatcher into figure builder

m-vo committed:
dispatch event when defining metadata

m-vo committed:
add and adjust tests

ausi commented on core-bundle/src/Image/Studio/FigureBuilder.php:
```diff
@@ -705,6 +710,11 @@ private function validatorAdapter()
         return $framework->getAdapter(Validator::class);
     }
 
+    private function eventDispatcher()
```
```suggestion
    private function eventDispatcher(): EventDispatcherInterface
```

Should we use the `Symfony\Contracts\EventDispatcher\EventDispatcherInterface` here?

m-vo replied:
sounds good

m-vo committed:
type hint with EventDispatcherInterface

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

leofeyer commented:
I would prefer `FigureMetadataEvent` or `FigureBuilderMetadataEvent` instead of `DefineMetadataEvent`, because the latter does not really say what kind of metadata it handles (there could be page metadata, news metadata, file metadata etc.). Or is this a general event that we can use for any metadata? Then I would just call it `MetadataEvent`.

Toflar commented:
It is `FileMetadataEvent`, I guess. The class name is `Contao\CoreBundle\File\Metadata`  :) (it represents meta data of `tl_files`).

m-vo commented:
We could also put the event in the File namespace. But that's not consistent with the other ones.

I have no strong preferences but would try to avoid having the FigureBuilder in the name.

leofeyer commented:
Then let's go with `FileMetadataEvent` but keep the class in the `Event` namespace.

m-vo committed:
rename event to FileMetadataEvent

m-vo committed:
add a simple unit test for the event class

m-vo commented:
Now renamed to `FileMetadataEvent` in 4babc1e59f1936c24f80c9496271e842af908220. I also added a small test for the event class in 1ea7904bd20701226529d385028863dcec2f9b7f.

leofeyer committed:
CS

leofeyer commented:
Thank you @m-vo.
